File: 14.Dynamic Programming/House Robber.py
from typing import *


class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        
        maxList = []
        maxList.append(nums[0])
        maxList.append(max(nums[0], nums[1]))

        for i in range(2, len(nums)):
            maxList.append(max(maxList[i - 1], maxList[i - 2] + nums[i]))

        return maxList[-1]

sol = Solution()
nums = [17, 11, 9, 4, 15, 12, 18, 8, 7, 8]

print(sol.rob(nums))

# import random
#
# list = []
#
# for i in range(10):
#     list.append(random.randrange(1, 20))
#
# print(list)

File: 14.Dynamic Programming/House Robber.java
package Week14_Dynamic_Programming;

import java.util.Arrays;
import java.util.OptionalInt;

public class Day1_House_Robber {
    public static int Wrongrob(int[] nums) {
    	int[] dp = new int[nums.length];
    	for(int i = 0 ; i < nums.length ; i++) {
    		dp[i] = nums[i];
    		for(int j = 0 ; j < i-1 ; j++) {
    			dp[i] = Math.max(dp[i], dp[j]+nums[i]);
    		}
    	}
    	int answer = Arrays.stream(dp).max().getAsInt();
//    	for(int d: dp) System.out.print(d+" ");
    	return answer;
    }
    public static int rob(int[] nums) {
    	int[] dp = new int[nums.length];
    	
    	if(nums.length == 1) return nums[0];
    	dp[0] = nums[0];
    	dp[1] = Math.max(dp[0], nums[1]);
    	
    	for(int i = 2 ; i < nums.length ; i++) {
    		dp[i] = Math.max(dp[i-1], dp[i-2]+nums[i]);
    	}
    	
    	int answer =dp[dp.length-1];
    	for(int d: dp) System.out.print(d+" ");
    	System.out.println();
    	return answer;
    }
	public static void main(String[] args) {
		int[] nums = { 1,2,3,1 };
//		int[] nums = {2,7,9,3,1};
//		int[] nums = { 2,1,1,2};
		System.out.println(rob(nums));
		

	}

}
